// @jsonforms/core/dist/test/test/util/renderer.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const ava_1 = require("ava");
const _ = require("lodash");
const util_1 = require("../../src/util");
const redux_mock_store_1 = require("redux-mock-store");
const actions_1 = require("../../src/actions");
const middlewares = [];
const mockStore = redux_mock_store_1.default(middlewares);
const hideRule = {
    effect: 'HIDE',
    condition: {
        type: 'LEAF',
        scope: '#/properties/firstName',
        expectedValue: 'Homer'
    }
};
const disableRule = {
    effect: 'DISABLE',
    condition: {
        type: 'LEAF',
        scope: '#/properties/firstName',
        expectedValue: 'Homer'
    }
};
const coreUISchema = {
    type: 'Control',
    scope: '#/properties/firstName',
};
const createState = uischema => ({
    jsonforms: {
        core: {
            schema: {
                type: 'object',
                properties: {
                    firstName: { type: 'string' },
                    lastName: { type: 'string' }
                }
            },
            data: {
                firstName: 'Homer'
            },
            uischema,
            errors: []
        },
        i18n: {
            locale: 'en-US'
        }
    }
});
ava_1.default('mapStateToControlProps - visible via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        visible: true,
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.true(props.visible);
});
ava_1.default('mapStateToControlProps - hidden via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        visible: false,
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.false(props.visible);
});
ava_1.default('mapStateToControlProps - hidden via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.false(props.visible);
});
ava_1.default('mapStateToControlProps - visible via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: hideRule });
    const ownProps = {
        uischema
    };
    const clonedState = _.cloneDeep(createState(uischema));
    clonedState.jsonforms.core.data.firstName = 'Lisa';
    const props = util_1.mapStateToControlProps(clonedState, ownProps);
    t.true(props.visible);
});
ava_1.default('mapStateToControlProps - enabled via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        enabled: true,
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.true(props.enabled);
});
ava_1.default('mapStateToControlProps - disabled via ownProps ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        enabled: false,
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.false(props.enabled);
});
ava_1.default('mapStateToControlProps - disabled via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        uischema
    };
    const props = util_1.mapStateToControlProps(createState(uischema), ownProps);
    t.false(props.enabled);
});
ava_1.default('mapStateToControlProps - enabled via state ', t => {
    const uischema = Object.assign({}, coreUISchema, { rule: disableRule });
    const ownProps = {
        uischema
    };
    const clonedState = _.cloneDeep(createState(uischema));
    clonedState.jsonforms.core.data.firstName = 'Lisa';
    const props = util_1.mapStateToControlProps(clonedState, ownProps);
    t.true(props.enabled);
});
ava_1.default('mapStateToControlProps - path', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.path, 'firstName');
});
ava_1.default('mapStateToControlProps - compose path with ownProps.path', t => {
    const ownProps = {
        uischema: coreUISchema,
        path: 'yo'
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.path, 'yo.firstName');
    t.is(props.parentPath, 'yo');
});
ava_1.default('mapStateToControlProps - derive label', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.label, 'First Name');
});
ava_1.default('mapStateToControlProps - derive label', t => {
    const ownProps = {
        uischema: Object.assign({}, coreUISchema, { label: {
                show: false
            } })
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.label, '');
});
ava_1.default('mapStateToControlProps - data', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.data, 'Homer');
});
ava_1.default('mapStateToControlProps - errors', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const clonedState = _.cloneDeep(createState(coreUISchema));
    clonedState.jsonforms.core.errors = [{
            dataPath: 'firstName',
            message: 'Duff beer'
        }];
    const props = util_1.mapStateToControlProps(clonedState, ownProps);
    t.is(props.errors[0], 'Duff beer');
});
ava_1.default('mapStateToControlProps - id', t => {
    const ownProps = {
        uischema: coreUISchema
    };
    const props = util_1.mapStateToControlProps(createState(coreUISchema), ownProps);
    t.is(props.id, '#/properties/firstName');
});
ava_1.default('mapDispatchToControlProps', t => {
    const store = mockStore(createState(coreUISchema));
    const props = util_1.mapDispatchToControlProps(store.dispatch);
    props.handleChange('foo', 42);
    const updateAction = _.head(store.getActions());
    t.is(updateAction.type, actions_1.UPDATE_DATA);
    t.is(updateAction.path, 'foo');
    t.is(updateAction.updater(), 42);
});
