// @jsonforms/core/dist/test/test/util/runtime.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
const ava_1 = require("ava");
const src_1 = require("../../src");
const runtime_1 = require("../../src/util/runtime");
ava_1.default('evalVisibility show valid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.SHOW,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'bar'
    };
    t.is(runtime_1.evalVisibility(uischema, data), true);
});
ava_1.default('evalVisibility show invalid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.SHOW,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'foobar'
    };
    t.deepEqual(runtime_1.evalVisibility(uischema, data), false);
});
ava_1.default('evalVisibility hide valid case', t => {
    const leafCondition = {
        'type': 'LEAF',
        'scope': '#/properties/ruleValue',
        'expectedValue': 'bar'
    };
    const uischema = {
        'type': 'Control',
        'scope': '#/properties/value',
        'rule': {
            'effect': src_1.RuleEffect.HIDE,
            'condition': leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'bar'
    };
    t.is(runtime_1.evalVisibility(uischema, data), false);
});
ava_1.default('evalVisibility hide invalid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.HIDE,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'foobar'
    };
    t.is(runtime_1.evalVisibility(uischema, data), true);
});
ava_1.default('evalEnablement enable valid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.ENABLE,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'bar'
    };
    t.is(runtime_1.evalEnablement(uischema, data), true);
});
ava_1.default('evalEnablement enable invalid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.ENABLE,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'foobar'
    };
    t.is(runtime_1.evalEnablement(uischema, data), false);
});
ava_1.default('evalEnablement disable valid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.DISABLE,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'bar'
    };
    t.is(runtime_1.evalEnablement(uischema, data), false);
});
ava_1.default('evalEnablement disable invalid case', t => {
    const leafCondition = {
        type: 'LEAF',
        scope: '#/properties/ruleValue',
        expectedValue: 'bar'
    };
    const uischema = {
        type: 'Control',
        scope: '#/properties/value',
        rule: {
            effect: src_1.RuleEffect.DISABLE,
            condition: leafCondition
        }
    };
    const data = {
        value: 'foo',
        ruleValue: 'foobar'
    };
    t.is(runtime_1.evalEnablement(uischema, data), true);
});
