// @jsonforms/react/dist/test/test/renderers/JsonForms.test.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
require("jsdom-global/register");
const installCE = require("document-register-element/pony");
installCE(global, 'force');
global.requestAnimationFrame = cb => setTimeout(cb, 0);
const React = require("react");
const ava_1 = require("ava");
const _ = require("lodash");
const redux_1 = require("redux");
const react_redux_1 = require("react-redux");
const core_1 = require("@jsonforms/core");
const TestUtils = require("react-dom/test-utils");
const src_1 = require("../../src");
exports.initJsonFormsStore = (_a) => {
    var { data, schema, uischema } = _a, props = __rest(_a, ["data", "schema", "uischema"]);
    return redux_1.createStore(redux_1.combineReducers({ jsonforms: core_1.jsonformsReducer() }), {
        jsonforms: Object.assign({ core: {
                data,
                schema,
                uischema
            } }, props)
    });
};
const CustomRenderer1 = () => (React.createElement("h1", null, "test"));
const CustomRenderer2 = () => (React.createElement("h2", null, "test"));
const CustomRenderer3 = () => (React.createElement("h3", null, "test"));
ava_1.test.beforeEach(t => {
    t.context.data = { foo: 'John Doe' };
    t.context.uischema = {
        type: 'Control',
        scope: '#/properties/foo'
    };
    t.context.schema = {
        type: 'object',
        properties: {
            foo: {
                type: 'string'
            }
        }
    };
});
ava_1.test('JsonForms renderer should report about missing renderer', t => {
    const data = { foo: 'John Doe' };
    const uischema = { type: 'Foo' };
    const schema = { type: 'object', properties: { foo: { type: 'string' } } };
    const store = exports.initJsonFormsStore({
        data,
        schema,
        uischema
    });
    const div = _.head(TestUtils.scryRenderedDOMComponentsWithTag(TestUtils.renderIntoDocument(React.createElement(react_redux_1.Provider, { store: store },
        React.createElement(src_1.JsonForms, { uischema: uischema, schema: schema }))), 'div'));
    t.is(div.textContent, 'No applicable renderer found.');
});
ava_1.test('JsonForms renderer should pick most applicable renderer', t => {
    const store = exports.initJsonFormsStore({
        data: t.context.data,
        schema: t.context.schema,
        uischema: t.context.uischema
    });
    store.dispatch(core_1.registerRenderer(() => 10, CustomRenderer1));
    store.dispatch(core_1.registerRenderer(() => 5, CustomRenderer1));
    const tree = TestUtils.renderIntoDocument(React.createElement(react_redux_1.Provider, { store: store },
        React.createElement(src_1.JsonForms, { uischema: t.context.uischema, schema: t.context.schema })));
    t.not(TestUtils.findRenderedDOMComponentWithTag(tree, 'h1'), undefined);
});
ava_1.test('JsonForms renderer should not consider any de-registered renderers', t => {
    const tester1 = () => 9;
    const tester2 = () => 8;
    const tester3 = () => 10;
    const store = exports.initJsonFormsStore({
        data: t.context.data,
        schema: t.context.schema,
        uischema: t.context.uischema
    });
    store.dispatch(core_1.registerRenderer(tester1, CustomRenderer1));
    store.dispatch(core_1.registerRenderer(tester2, CustomRenderer2));
    store.dispatch(core_1.registerRenderer(tester3, CustomRenderer3));
    store.dispatch(core_1.unregisterRenderer(tester3, CustomRenderer2));
    const tree = TestUtils.renderIntoDocument(React.createElement(react_redux_1.Provider, { store: store },
        React.createElement(src_1.JsonForms, { uischema: t.context.uischema, schema: t.context.schema })));
    t.not(TestUtils.findRenderedDOMComponentWithTag(tree, 'h1'), undefined);
});
ava_1.test('deregister an unregistered renderer should be a no-op', t => {
    const store = exports.initJsonFormsStore({
        data: t.context.data,
        schema: t.context.schema,
        uischema: t.context.uischema
    });
    store.dispatch(core_1.registerRenderer(() => 10, CustomRenderer1));
    store.dispatch(core_1.registerRenderer(() => 5, CustomRenderer2));
    const tester = () => 10;
    const nrOfRenderers = store.getState().jsonforms.renderers.length;
    store.dispatch(core_1.unregisterRenderer(tester, CustomRenderer3));
    t.is(store.getState().jsonforms.renderers.length, nrOfRenderers);
});
