// @jsonforms/core/lib/testers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var _ = window.lodashs;
var resolvers_1 = require("../util/resolvers");
/**
 * Constant that indicates that a tester is not capable of handling
 * a combination of schema/data.
 * @type {number}
 */
exports.NOT_APPLICABLE = -1;
exports.isControl = function (uischema) {
    return !_.isEmpty(uischema) && uischema.scope !== undefined && uischema.scope !== undefined;
};
/**
 * Only applicable for Controls.
 *
 * This function checks whether the given UI schema is of type Control
 * and if so, resolves the sub-schema referenced by the control and applies
 * the given predicate
 *
 * @param {(JsonSchema) => boolean} predicate the predicate that should be
 *        applied to the resolved sub-schema
 */
exports.schemaMatches = function (predicate) {
    return function (uischema, schema) {
        if (_.isEmpty(uischema) || !exports.isControl(uischema)) {
            return false;
        }
        var schemaPath = uischema.scope;
        if (_.isEmpty(schemaPath)) {
            return false;
        }
        var currentDataSchema = resolvers_1.resolveSchema(schema, schemaPath);
        while (!_.isEmpty(currentDataSchema) && !_.isEmpty(currentDataSchema.$ref)) {
            currentDataSchema = resolvers_1.resolveSchema(schema, currentDataSchema.$ref);
        }
        if (currentDataSchema === undefined) {
            return false;
        }
        return predicate(currentDataSchema);
    };
};
exports.schemaSubPathMatches = function (subPath, predicate) {
    return function (uischema, schema) {
        if (_.isEmpty(uischema) || !exports.isControl(uischema)) {
            return false;
        }
        var schemaPath = uischema.scope;
        if (_.isEmpty(schemaPath)) {
            return false;
        }
        var currentDataSchema = resolvers_1.resolveSchema(schema, schemaPath + "/" + subPath);
        while (!_.isEmpty(currentDataSchema.$ref)) {
            currentDataSchema = resolvers_1.resolveSchema(schema, currentDataSchema.$ref);
        }
        if (currentDataSchema === undefined) {
            return false;
        }
        return predicate(currentDataSchema);
    };
};
/**
 * Only applicable for Controls.
 *
 * This function checks whether the given UI schema is of type Control
 * and if so, resolves the sub-schema referenced by the control and checks
 * whether the type of the sub-schema matches the expected one.
 *
 * @param {string} expectedType the expected type of the resolved sub-schema
 */
exports.schemaTypeIs = function (expectedType) { return exports.schemaMatches(function (schema) {
    return !_.isEmpty(schema) && schema.type === expectedType;
}); };
/**
 * Only applicable for Controls.
 *
 * This function checks whether the given UI schema is of type Control
 * and if so, resolves the sub-schema referenced by the control and checks
 * whether the format of the sub-schema matches the expected one.
 *
 * @param {string} expectedFormat the expected format of the resolved sub-schema
 */
exports.formatIs = function (expectedFormat) { return exports.schemaMatches(function (schema) {
    return !_.isEmpty(schema)
        && schema.format === expectedFormat
        && schema.type === 'string';
}); };
/**
 * Checks whether the given UI schema has the expected type.
 *
 * @param {string} expected the expected UI schema type
 */
exports.uiTypeIs = function (expected) {
    return function (uischema) {
        return !_.isEmpty(uischema) && uischema.type === expected;
    };
};
/**
 * Checks whether the given UI schema has an option with the given
 * name and whether it has the expected value. If no options property
 * is set, returns false.
 *
 * @param {string} optionName the name of the option to check
 * @param {any} optionValue the expected value of the option
 */
exports.optionIs = function (optionName, optionValue) {
    return function (uischema) {
        var options = uischema.options;
        return !_.isEmpty(options) && options[optionName] === optionValue;
    };
};
/**
 * Only applicable for Controls.
 *
 * Checks whether the scope of a control ends with the expected string.
 *
 * @param {string} expected the expected ending of the reference
 */
exports.scopeEndsWith = function (expected) {
    return function (uischema) {
        if (_.isEmpty(expected) || !exports.isControl(uischema)) {
            return false;
        }
        return _.endsWith(uischema.scope, expected);
    };
};
/**
 * Only applicable for Controls.
 *
 * Checks whether the last segment of the scope matches the expected string.
 *
 * @param {string} expected the expected ending of the reference
 */
exports.scopeEndIs = function (expected) {
    return function (uischema) {
        if (_.isEmpty(expected) || !exports.isControl(uischema)) {
            return false;
        }
        var schemaPath = uischema.scope;
        return !_.isEmpty(schemaPath) && _.last(schemaPath.split('/')) === expected;
    };
};
/**
 * A tester that allow composing other testers by && them.
 *
 * @param {Array<Tester>} testers the testers to be composed
 */
exports.and = function () {
    var testers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        testers[_i] = arguments[_i];
    }
    return function (uischema, schema) {
        return testers.reduce(function (acc, tester) { return acc && tester(uischema, schema); }, true);
    };
};
/**
 * A tester that allow composing other testers by || them.
 *
 * @param {Array<Tester>} testers the testers to be composed
 */
exports.or = function () {
    var testers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        testers[_i] = arguments[_i];
    }
    return function (uischema, schema) {
        return testers.reduce(function (acc, tester) { return acc || tester(uischema, schema); }, false);
    };
};
/**
 * Create a ranked tester that will associate a number with a given tester, if the
 * latter returns true.
 *
 * @param {number} rank the rank to be returned in case the tester returns true
 * @param {Tester} tester a tester
 */
exports.rankWith = function (rank, tester) {
    return function (uischema, schema) {
        if (tester(uischema, schema)) {
            return rank;
        }
        return exports.NOT_APPLICABLE;
    };
};
exports.withIncreasedRank = function (by, rankedTester) {
    return function (uischema, schema) {
        var rank = rankedTester(uischema, schema);
        if (rank === exports.NOT_APPLICABLE) {
            return exports.NOT_APPLICABLE;
        }
        return rank + by;
    };
};
/**
 * Default tester for boolean.
 * @type {RankedTester}
 */
exports.isBooleanControl = exports.and(exports.uiTypeIs('Control'), exports.schemaTypeIs('boolean'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * has a 'date' format.
 * @type {Tester}
 */
exports.isDateControl = exports.and(exports.uiTypeIs('Control'), exports.formatIs('date'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * has an enum.
 * @type {Tester}
 */
exports.isEnumControl = exports.and(exports.uiTypeIs('Control'), exports.schemaMatches(function (schema) { return schema.hasOwnProperty('enum'); }));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * is of type integer
 * @type {Tester}
 */
exports.isIntegerControl = exports.and(exports.uiTypeIs('Control'), exports.schemaTypeIs('integer'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * is of type number
 * @type {Tester}
 */
exports.isNumberControl = exports.and(exports.uiTypeIs('Control'), exports.schemaTypeIs('number'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * is of type string
 * @type {Tester}
 */
exports.isStringControl = exports.and(exports.uiTypeIs('Control'), exports.schemaTypeIs('string'));
/**
 * Tests whether the given UI schema is of type Control and if is has
 * a 'multi' option.
 * @type {Tester}
 */
exports.isMultiLineControl = exports.and(exports.uiTypeIs('Control'), exports.optionIs('multi', true));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * has a 'time' format.
 * @type {Tester}
 */
exports.isTimeControl = exports.and(exports.uiTypeIs('Control'), exports.formatIs('time'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * has a 'date-time' format.
 * @type {Tester}
 */
exports.isDateTimeControl = exports.and(exports.uiTypeIs('Control'), exports.formatIs('date-time'));
/**
 * Tests whether the given UI schema is of type Control and if the schema
 * is an array of objects.
 * @type {Tester}
 */
exports.isArrayObjectControl = exports.and(exports.uiTypeIs('Control'), exports.schemaMatches(function (schema) {
    return !_.isEmpty(schema)
        && schema.type === 'array'
        && !_.isEmpty(schema.items)
        && !Array.isArray(schema.items);
} // we don't care about tuples
), exports.schemaSubPathMatches('items', function (schema) {
    return schema.type === 'object';
}));
/**
 * Tests whether a given UI schema is of type Control and
 * whether the schema defines a numerical range.
 * Furthermore it checks that the Control has the option 'slider' set to 'true'.
 * @type {Tester}
 */
exports.isRangeControl = exports.and(exports.uiTypeIs('Control'), exports.or(exports.schemaTypeIs('number'), exports.schemaTypeIs('integer')), exports.schemaMatches(function (schema) {
    return schema.hasOwnProperty('maximum') &&
        schema.hasOwnProperty('minimum') &&
        schema.hasOwnProperty('default');
}));
/**
 * Tests whether the given UI schema is of type Control, if the schema
 * is of type string and has option format
 * @type {Tester}
 */
exports.isNumberFormatControl = exports.and(exports.uiTypeIs('Control'), exports.schemaTypeIs('integer'), exports.optionIs('format', true));
