// @jsonforms/material-renderers/lib/layouts/MaterialCategorizationLayout.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var React = require("react");
var AppBar_1 = require("material-ui/AppBar");
var Tabs_1 = require("material-ui/Tabs");
var core_1 = require("@jsonforms/core");
var react_1 = require("@jsonforms/react");
var layout_1 = require("../util/layout");
var isSingleLevelCategorization = core_1.and(core_1.uiTypeIs('Categorization'), function (uischema) {
    var categorization = uischema;
    return categorization.elements.reduce(function (acc, e) { return acc && e.type === 'Category'; }, true);
});
exports.materialCategorizationTester = core_1.rankWith(1, isSingleLevelCategorization);
var MaterialCategorizationLayoutRenderer = /** @class */ (function (_super) {
    __extends(MaterialCategorizationLayoutRenderer, _super);
    function MaterialCategorizationLayoutRenderer(props) {
        var _this = _super.call(this, props) || this;
        _this.handleChange = function (_event, value) {
            _this.setState({ value: value });
        };
        _this.state = {
            value: 0
        };
        return _this;
    }
    MaterialCategorizationLayoutRenderer.prototype.render = function () {
        var _a = this.props, uischema = _a.uischema, schema = _a.schema, path = _a.path, visible = _a.visible;
        var value = this.state.value;
        var categorization = uischema;
        var childProps = {
            elements: categorization.elements[value].elements,
            schema: schema,
            path: path,
            direction: 'column',
            visible: visible
        };
        var style = { marginBottom: '10px' };
        if (!visible) {
            style.display = 'none';
        }
        return (React.createElement("div", { style: style },
            React.createElement(AppBar_1.default, { position: 'static' },
                React.createElement(Tabs_1.default, { value: value, onChange: this.handleChange }, categorization.elements.map(function (e, idx) { return React.createElement(Tabs_1.Tab, { key: idx, label: e.label }); }))),
            React.createElement("div", { style: { marginTop: '0.5em' } },
                React.createElement(layout_1.MaterialLayoutRenderer, __assign({}, childProps)))));
    };
    return MaterialCategorizationLayoutRenderer;
}(react_1.RendererComponent));
exports.MaterialCategorizationLayoutRenderer = MaterialCategorizationLayoutRenderer;
exports.default = react_1.connectToJsonForms(core_1.mapStateToLayoutProps)(MaterialCategorizationLayoutRenderer);
