// @jsonforms/core/lib/util/field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var _ = window.lodashs;
var reducers_1 = require("../reducers");
var util_1 = require("../util");
var renderer_1 = require("./renderer");
exports.mapStateToDispatchFieldProps = function (state) { return ({
    fields: state.jsonforms.fields || []
}); };
/**
 * Map state to field props.
 *
 * @param state JSONForms state tree
 * @param ownProps any own props
 * @returns {StatePropsOfField} state props of a field
 */
exports.mapStateToFieldProps = function (state, ownProps) {
    var path = util_1.composeWithUi(ownProps.uischema, ownProps.path);
    var visible = _.has(ownProps, 'visible') ? ownProps.visible : util_1.isVisible(ownProps, state);
    var enabled = _.has(ownProps, 'enabled') ? ownProps.enabled : util_1.isEnabled(ownProps, state);
    var errors = reducers_1.getErrorAt(path)(state).map(function (error) { return error.message; });
    var isValid = _.isEmpty(errors);
    var controlElement = ownProps.uischema;
    var id = controlElement.scope || '';
    var inputClassName = ['validate'].concat(isValid ? 'valid' : 'invalid');
    var defaultConfig = _.cloneDeep(reducers_1.getConfig(state));
    var config = _.merge(defaultConfig, ownProps.uischema.options);
    return {
        data: util_1.Resolve.data(reducers_1.getData(state), path),
        className: inputClassName.join(' '),
        visible: visible,
        enabled: enabled,
        id: id,
        path: path,
        isValid: isValid,
        scopedSchema: util_1.Resolve.schema(ownProps.schema, controlElement.scope),
        uischema: ownProps.uischema,
        schema: ownProps.schema,
        config: config
    };
};
/**
 * Synonym for mapDispatchToControlProps.
 *
 * @type {(dispatch) => {handleChange(path, value): void}}
 */
exports.mapDispatchToFieldProps = renderer_1.mapDispatchToControlProps;
