// @jsonforms/material-renderers/lib/controls/MaterialDateControl.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var React = require("react");
var _ = require("lodash");
var core_1 = require("@jsonforms/core");
var react_1 = require("@jsonforms/react");
var material_ui_pickers_1 = require("material-ui-pickers");
var KeyboardArrowLeft_1 = require("material-ui-icons/KeyboardArrowLeft");
var KeyboardArrowRight_1 = require("material-ui-icons/KeyboardArrowRight");
var moment = require("moment");
var MaterialDateControl = /** @class */ (function (_super) {
    __extends(MaterialDateControl, _super);
    function MaterialDateControl() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MaterialDateControl.prototype.render = function () {
        var _a = this.props, description = _a.description, id = _a.id, errors = _a.errors, label = _a.label, uischema = _a.uischema, visible = _a.visible, enabled = _a.enabled, required = _a.required, path = _a.path, handleChange = _a.handleChange, data = _a.data, momentLocale = _a.momentLocale;
        var isValid = errors.length === 0;
        var trim = uischema.options && uischema.options.trim;
        var showDescription = !core_1.isDescriptionHidden(visible, description, this.state.isFocused);
        var style = {};
        if (!visible) {
            style = { display: 'none' };
        }
        var inputProps = {};
        var localeDateTimeFormat = momentLocale ? "" + momentLocale.localeData().longDateFormat('L')
            : 'YYYY-MM-DD';
        var labelText;
        var labelCancel;
        var labelClear;
        if (core_1.isPlainLabel(label)) {
            labelText = label;
            labelCancel = 'Cancel';
            labelClear = 'Clear';
        }
        else {
            labelText = label.default;
            labelCancel = _.startsWith(label.cancel, '%') ? 'Cancel' : label.cancel;
            labelClear = _.startsWith(label.clear, '%') ? 'Clear' : label.clear;
        }
        return (React.createElement(material_ui_pickers_1.DatePicker, { id: id, label: core_1.computeLabel(labelText, required), error: !isValid, style: style, fullWidth: !trim, helperText: !isValid ? errors : showDescription ? description : null, InputLabelProps: { shrink: true }, value: data || null, onChange: function (datetime) {
                return handleChange(path, datetime ? moment(datetime).format('YYYY-MM-DD') : '');
            }, format: localeDateTimeFormat, clearable: true, disabled: !enabled, autoFocus: uischema.options && uischema.options.focus, onClear: function () { return handleChange(path, ''); }, onFocus: this.onFocus, onBlur: this.onBlur, cancelLabel: labelCancel, clearLabel: labelClear, leftArrowIcon: React.createElement(KeyboardArrowLeft_1.default, null), rightArrowIcon: React.createElement(KeyboardArrowRight_1.default, null), InputProps: inputProps }));
    };
    return MaterialDateControl;
}(react_1.Control));
exports.MaterialDateControl = MaterialDateControl;
exports.addLabelProps = function (mapStateToProps) {
    return function (state, ownProps) {
        var stateProps = mapStateToProps(state, ownProps);
        return __assign({}, stateProps, { label: {
                default: stateProps.label,
                cancel: '%cancel',
                clear: '%clear'
            } });
    };
};
exports.materialDateControlTester = core_1.rankWith(4, core_1.isDateControl);
exports.default = react_1.connectToJsonForms(exports.addLabelProps(core_1.mapStateToControlProps), core_1.mapDispatchToControlProps)(MaterialDateControl);
