// @jsonforms/core/dist/test/src/actions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const generators_1 = require("../generators");
const NAMESPACE = 'jsonforms';
exports.INIT = `${NAMESPACE}/INIT`;
exports.UPDATE_DATA = 'jsonforms/UPDATE';
exports.VALIDATE = `${NAMESPACE}/VALIDATE`;
exports.ADD_RENDERER = `${NAMESPACE}/ADD_RENDERER`;
exports.REMOVE_RENDERER = `${NAMESPACE}/REMOVE_RENDERER`;
exports.ADD_FIELD = `${NAMESPACE}/ADD_FIELD`;
exports.REMOVE_FIELD = `${NAMESPACE}/REMOVE_FIELD`;
exports.SET_CONFIG = `${NAMESPACE}/SET_CONFIG`;
exports.init = (data, schema = generators_1.generateJsonSchema(data), uischema = generators_1.generateDefaultUISchema(schema)) => ({
    type: exports.INIT,
    data,
    schema,
    uischema
});
exports.update = (path, updater) => ({
    type: exports.UPDATE_DATA,
    path,
    updater
});
exports.registerRenderer = (tester, renderer) => ({
    type: exports.ADD_RENDERER,
    tester,
    renderer
});
exports.registerField = (tester, field) => ({
    type: exports.ADD_FIELD,
    tester,
    field
});
exports.unregisterField = (tester, field) => ({
    type: exports.REMOVE_FIELD,
    tester,
    field
});
exports.unregisterRenderer = (tester, renderer) => ({
    type: exports.REMOVE_RENDERER,
    tester,
    renderer
});
exports.setConfig = config => dispatch => {
    dispatch({
        type: exports.SET_CONFIG,
        config,
    });
};
