// @jsonforms/core/lib/actions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var generators_1 = require("../generators");
var NAMESPACE = 'jsonforms';
exports.INIT = NAMESPACE + "/INIT";
exports.UPDATE_DATA = 'jsonforms/UPDATE';
exports.VALIDATE = NAMESPACE + "/VALIDATE";
exports.ADD_RENDERER = NAMESPACE + "/ADD_RENDERER";
exports.REMOVE_RENDERER = NAMESPACE + "/REMOVE_RENDERER";
exports.ADD_FIELD = NAMESPACE + "/ADD_FIELD";
exports.REMOVE_FIELD = NAMESPACE + "/REMOVE_FIELD";
exports.SET_CONFIG = NAMESPACE + "/SET_CONFIG";
exports.init = function (data, schema, uischema) {
    if (schema === void 0) { schema = generators_1.generateJsonSchema(data); }
    if (uischema === void 0) { uischema = generators_1.generateDefaultUISchema(schema); }
    return ({
        type: exports.INIT,
        data: data,
        schema: schema,
        uischema: uischema
    });
};
exports.update = function (path, updater) { return ({
    type: exports.UPDATE_DATA,
    path: path,
    updater: updater
}); };
exports.registerRenderer = function (tester, renderer) { return ({
    type: exports.ADD_RENDERER,
    tester: tester,
    renderer: renderer
}); };
exports.registerField = function (tester, field) { return ({
    type: exports.ADD_FIELD,
    tester: tester,
    field: field
}); };
exports.unregisterField = function (tester, field) { return ({
    type: exports.REMOVE_FIELD,
    tester: tester,
    field: field
}); };
exports.unregisterRenderer = function (tester, renderer) { return ({
    type: exports.REMOVE_RENDERER,
    tester: tester,
    renderer: renderer
}); };
exports.setConfig = function (config) { return function (dispatch) {
    dispatch({
        type: exports.SET_CONFIG,
        config: config,
    });
}; };
