// @jsonforms/material-renderers/lib/complex/MaterialTableControl.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
/*
  The MIT License
  
  Copyright (c) 2018 EclipseSource Munich
  https://github.com/eclipsesource/jsonforms
  
  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:
  
  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
*/
var React = require("react");
var _ = require("lodash");
var Checkbox_1 = require("material-ui/Checkbox");
var Grid_1 = require("material-ui/Grid");
var Table_1 = require("material-ui/Table");
var core_1 = require("@jsonforms/core");
var react_1 = require("@jsonforms/react");
var ValidationIcon_1 = require("./ValidationIcon");
var generateCells = function (Cell, scopedSchema, rowPath, cellErrors) {
    return Object.keys(scopedSchema.properties)
        .filter(function (prop) { return scopedSchema.properties[prop].type !== 'array'; })
        .map(function (prop) {
        var cellPath = core_1.Paths.compose(rowPath, prop);
        var props = {
            cellProperty: prop,
            scopedSchema: scopedSchema,
            rowPath: rowPath,
            cellPath: cellPath,
            errors: cellErrors
        };
        return React.createElement(Cell, __assign({ key: cellPath }, props));
    });
};
var EmptyTable = function () { return (React.createElement(Table_1.TableRow, null,
    React.createElement(Table_1.TableCell, null, "No data"))); };
var TableHeaderCell = function (_a) {
    var cellProperty = _a.cellProperty;
    return React.createElement(Table_1.TableCell, null, _.capitalize(cellProperty));
};
var TableContentCell = function (_a) {
    var rowPath = _a.rowPath, cellProperty = _a.cellProperty, cellPath = _a.cellPath, errors = _a.errors, scopedSchema = _a.scopedSchema;
    var cellErrors = errors
        .filter(function (error) { return error.dataPath === cellPath; })
        .map(function (error) { return error.message; });
    var createControlElement = function (key) { return ({
        type: 'Control',
        label: false,
        scope: "#/properties/" + key
    }); };
    return (React.createElement(Table_1.TableCell, null,
        React.createElement(Grid_1.default, { container: true, alignItems: 'center', justify: 'center', spacing: 0 },
            React.createElement(Grid_1.default, { item: true, xs: 1, hidden: { smUp: cellErrors.length === 0 } },
                React.createElement(ValidationIcon_1.ValidationIcon, { id: "tooltip-" + cellPath, errorMessages: cellErrors })),
            React.createElement(Grid_1.default, { item: true, xs: true },
                React.createElement(react_1.DispatchField, { schema: scopedSchema, uischema: createControlElement(cellProperty), path: rowPath })))));
};
var TableWithContent = function (tableProps) {
    var data = tableProps.data, path = tableProps.path, scopedSchema = tableProps.scopedSchema, childErrors = tableProps.childErrors, select = tableProps.select, isSelected = tableProps.isSelected;
    return data.map(function (_child, index) {
        var childPath = core_1.Paths.compose(path, "" + index);
        var selected = isSelected(index);
        return (React.createElement(Table_1.TableRow, { key: childPath, hover: true, selected: selected },
            React.createElement(Table_1.TableCell, { padding: 'checkbox' },
                React.createElement(Checkbox_1.default, { checked: selected, onChange: function (e) { return select(e, index); } })),
            generateCells(TableContentCell, scopedSchema, childPath, childErrors)));
    });
};
exports.MaterialTableControl = function (props) {
    var data = props.data, path = props.path, scopedSchema = props.scopedSchema, numSelected = props.numSelected, selectAll = props.selectAll;
    var isEmptyTable = !data || !Array.isArray(data) || data.length === 0;
    var rowCount = data ? data.length : 0;
    return (React.createElement(Table_1.default, null,
        React.createElement(Table_1.TableHead, null,
            React.createElement(Table_1.TableRow, null,
                React.createElement(Table_1.TableCell, { padding: 'checkbox', style: { width: '1em' } },
                    React.createElement(Checkbox_1.default, { indeterminate: numSelected > 0 && numSelected < rowCount, checked: numSelected === rowCount, onChange: selectAll })),
                generateCells(TableHeaderCell, scopedSchema, path))),
        React.createElement(Table_1.TableBody, null, isEmptyTable ? React.createElement(EmptyTable, null) : React.createElement(TableWithContent, __assign({}, props)))));
};
